python 2 to 3 porting CI

jdejoya17 opened the pull request:
- added porting CI to check for python 2 to 3 issues
- porting issue fixes to the modules

pvtorres commented on roles/inetd/defaults/main.yml:
```diff
@@ -2,4 +2,4 @@
 services: []
 state: enabled
-backup: no
+backup: false
```
Why this was changed to false?

jdejoya17 replied:
there is no change in behavior here, it is just that the yaml linting followed by ansible for boolean like values, they prefer that is should be either `false` and `true` in this particular format/letter case.

pvtorres commented on roles/power_aix_bootstrap/tasks/python_install.yml:
```diff
@@ -1,33 +1,34 @@
+- name: Check for existence of yum
+  raw: "rpm -q --quiet yum && echo true || echo false"
+  register: yum_exists
+  changed_when: false
```
Why does the changed_when as added here?

jdejoya17 replied:
the `raw` module will always report `changed` even tho in this particular task nothing will every changes, it is merely querying for the existence of yum in the lpar. hence, it is not appropriate to report `changed` when nothing will ever change. so the role linting CI flagged this and this is how you fix it

pvtorres commented on roles/power_aix_bootstrap/tasks/python_install.yml:
```diff
@@ -1,33 +1,34 @@
+- name: Install/ Update python and requisite rpms
+  yum:
+    name: python
+    state: present
```
Why changed this to present as the previous one was set to latest. If package is present, it will be updated to the latest.

jdejoya17 replied:
the role linting CI flags this and recommends to use `present` instead of `latest`.

i think the rationale for this is `present` is used, then if something was already installed then yum module will do nothing else.
if `latest` is used, then in addition to checking if something is already installed it will also update it to the latest version if one is available.
the whole point in this is that if something was already installed and is working, then it should be kept as is unless explicitly specified to update. the reason for it is if the `latest` version has for some reason breaks functionality for anything that depends on it or any of the dependency is not compatible anymore, then it will affect the system.
so that is why they recommend using `present` if it is not already installed.

pvtorres replied:
Is there a way to ignore this error? In the current dnf bootstrap is require to update the package if it is present. And the yum module is allowing to have this functionality in there.

jdejoya17 replied:
updated in a different PR

pvtorres commented on roles/power_aix_bootstrap/tasks/yum_install.yml:
```diff
@@ -1,98 +1,105 @@
+- name: Check for existence of yum
+  raw: "rpm -q --quiet yum && echo true || echo false"
+  register: yum_exists
+  changed_when: false
```
Why this was added here?

jdejoya17 replied:
if you are asking on the `changed_when: false` then it is the same as the answer from the previous comment

pvtorres commented on roles/power_aix_bootstrap/tasks/yum_install.yml:
```diff
@@ -1,98 +1,105 @@
-    when: scp_result.rc == 0
+- name: Restore yum bundle content
+  raw: "{{ target_dir }}/yum_installer.sh {{ target_dir }}"
+  failed_when: false
```
We need to investigate the reason the playbook was set to ignore the errors

jdejoya17 replied:
i am not sure when about the reason why this task will fail, but it seems that in the original playbook it didn't matter if this task fails or not (i assume, it is related to trying to run this command, but ends up not being needed, so it returns error because there is nothing to do - like having an error when you use `rm` command to a file that does not exist)

so in the role linting CI, ansible prefers to used `failed_when` for cases where you ignore errors instead

pvtorres committed:
- add python3 porting test to CI
- fixed porting issues
- update sanity workflow to make sure a specific
pylint version is used where --py3k flag is supported

MGomathi commented on plugins/modules/nim_vios_alt_disk.py:
```diff
@@ -874,7 +874,7 @@ def check_rootvg(module, vios):
         return vg_info
 
-    if len(copy_dict.keys()) > 1:
+    if len(list(copy_dict.keys())) > 1:
```
At line no:1153:
nb_copies = len(copies_h.keys())
Does this also need len(list(copies_h.keys())

jdejoya17 replied:
@MGomathi
it looks like the linter will only look at the first level of dict. not sure if that is how they intended it.
i was testing with a dictionary that has a dictionary inside.
if i use the .keys() on the root dictionary, pylint will flag it. however if i use the .keys() in the dictinary inside the dictionary, then pylint will not flag it. it is strange. but it won't make any harm as the line you provided they were only interested in figuring out the number of keys. the iterator (from py3 .keys()) and the list (from py2 .keys()) will have the same length.

i could ignore this instead. what do you think?

pvtorres replied:
Joseph and I discussed about this. We are going to leave it as it is right now. The result will not change with our without the changes and pylint  is not complaining.

jdejoya17 committed:
temporarily ignore cStringIO module import. The module cStringIO
must be replaced with the io module to make sure it is compatible
with python3

jdejoya17 commented on roles/power_aix_vioshc/files/vioshc.py:
```diff
@@ -26,7 +27,7 @@
 import xml.etree.cElementTree as ET
 import socket
-import cStringIO
+import cStringIO # pylint: disable=bad-python3-import
```
this is how you ignore pylint errors/warnings
